### PR TITLE
Model scripting

### PR DESCRIPTION
**Pull request by sethalves**
- new javascript call: Model.appendMeshes which can combine multiple meshes into one
- new javascript call: Model.transformMesh which multiplies a matrix into all the vertices in a mesh and returns a new mesh with the results

### CODE REVIEW

**Comment by sethalves** [comment]
to test:
- find a domain where you are allowed to write to the asset-server
- run http://headache.hungry.com/~seth/hifi/voxel-to-mesh-test/voxel-to-mesh-test.js which is reproduced here:
```javascript

"use strict";

/* global Entities, Script, Tablet, Vec3, MyAvatar, Model, Assets */

(function() { // BEGIN LOCAL_SCOPE

    var tablet = Tablet.getTablet("com.highfidelity.interface.tablet.system");
    var button = tablet.addButton({
        icon: "http://headache.hungry.com/~seth/hifi/voxel-to-mesh-test/voxel-to-mesh-test.svg",
        text: "VtoM Test",
        sortOrder: 15
    });

    function retryingVoxelsToMesh(entityID, callback) {
        // if the polyvox mesh isn't yet computed, Entities.voxelsToMesh will start the process and call-back
        // with success set to false.
        Entities.voxelsToMesh(entityID, function(mesh, success) {
            if (success) {
                callback(mesh, true);
            } else {
                Script.setTimeout(function() {
                    retryingVoxelsToMesh(entityID, callback);
                }, 50);
            }
        });
    }

    function onClicked() {
        var nth = Math.floor((Math.random() * 1000) + 1);

        var position = Vec3.sum(MyAvatar.position, Vec3.multiplyQbyV(MyAvatar.orientation, {x: 0, y: 0.5, z: -3.5}));
        var toTheSide = Vec3.multiplyQbyV(MyAvatar.orientation, {x: 1.2, y: 0, z: 0});
        var fileName = "/model-scripting-test-" + nth + ".obj";

        var voxelID0 = Entities.addEntity({
            type: "PolyVox",
            dimensions: { x: 2, y: 2, z: 2 },
            voxelVolumeSiez: { x: 32, y: 32, z: 32 },
            position: Vec3.sum(position, toTheSide),
            rotation: MyAvatar.orientation,
            name: "polyvox mesh extract test 0",
            lifetime: 600,
            dynamic: false,
            collisionless: true,
            // voxelSurfaceStyle: 2 // cubic
            voxelSurfaceStyle: 3 // edged marching-cubes
        });

        var voxelID1 = Entities.addEntity({
            type: "PolyVox",
            dimensions: { x: 2, y: 2, z: 2 },
            voxelVolumeSiez: { x: 32, y: 32, z: 32 },
            position: Vec3.subtract(position, toTheSide),
            rotation: MyAvatar.orientation,
            name: "polyvox mesh extract test 1",
            lifetime: 600,
            dynamic: false,
            collisionless: true,
            // voxelSurfaceStyle: 2 // edged cubic
            voxelSurfaceStyle: 3 // edged marching-cubes
        });


        Entities.setVoxelSphere(voxelID0, position, 1, 255);
        Entities.setVoxelSphere(voxelID0, position, 0.5, 0);
        Entities.setVoxelSphere(voxelID1, position, 1, 255);

        retryingVoxelsToMesh(voxelID0, function(mesh0) {
            retryingVoxelsToMesh(voxelID1, function(mesh1) {
                var mesh0T = Model.transformMesh(Entities.getEntityTransform(voxelID0), mesh0);
                var mesh1T = Model.transformMesh(Entities.getEntityTransform(voxelID1), mesh1);

                // this way causes the resulting obj to have two sub-parts
                // var objData = Model.meshToOBJ([mesh0T, mesh1T]);

                // this way it's just one mesh-part
                var combinedMeshes = Model.appendMeshes([mesh0T, mesh1T]);
                var objData = Model.meshToOBJ([combinedMeshes]);


                Assets.uploadData(objData, function(url, hash) {
                    Assets.setMapping(fileName, hash, function() {
                        Entities.addEntity({
                            type: "Model",
                            modelURL: "atp:" + fileName,
                            // dimensions: { x: 2, y: 2, z: 2 },
                            position: Vec3.sum(position, { x: 0, y: 2, z: 0 }),
                            // rotation: MyAvatar.orientation,
                            name: "polyvox mesh extract test model",
                            lifetime: 600,
                            dynamic: false,
                            collisionless: true
                        });
                    });
                });
            });
        });
    }

    function cleanup() {
        button.clicked.disconnect(onClicked);
        tablet.removeButton(button);
    }

    button.clicked.connect(onClicked);
    Script.scriptEnding.connect(cleanup);
}()); // END LOCAL_SCOPE
```
- click on the added button
- look for two polyvox entities, each with roughly a half circle, separated by a small gap
- look for a single obj model entity placed above the polyvox entities and shaped like the polyvox entities combined.
- normals will be incorrect.  obj models with no material information will change colors and wink in and out as you move.  These problems aren't directly related to these changes.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 35c38b81300a136369c08a7176ef3ad68e5d6cb3 available at: https://gist.github.com/b474f3429ca2743df186394b673dbdf7<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9946-35c38b8.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9946-35c38b8.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: b876d22e46f21bd4c6e87eec2b8c6675255975cc available at: https://gist.github.com/b474f3429ca2743df186394b673dbdf7<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9946-b876d22.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9946-b876d22.dmg

**Comment by Nex-Pro** [comment]
Test in progress.

**Comment by Nex-Pro** [comment]
#9946 QA.

Testing platform: Intel workstation.
OS: Win10 pro 64
CPU: i7 4.0 ghz
RAM: 16gb
GPU: Nvidia GTX1070 8gb VRam
HMD: Htc Vive.

Specific tests:

	click on the added button: **DONE**

	look for two polyvox entities, each with roughly a half circle, separated by a small gap: **SUCCESS**
	look for a single obj model entity placed above the polyvox entities and shaped like the polyvox entities combined: **SUCCESS**
	normals will be incorrect. obj models with no material information will change colors and wink in and out as you move. These problems aren't directly related to these changes: **BEHAVIOUR AS DESCRIBED**

### PASS for specifics ( read observations below, look at images )


Smoke test:

Before running smoke test I made a bunch of these entities and selected the various surface extraction modes.

Also ran the script changing the surface extraction mode.

## QA Test
**QA Tester**: Piper.Peppercorn
**localhost 'createTestBlocks.js' Result**: SUCCESS
**localhost ATP Upload + Edit Result**:  SUCCESS
**Echo Server Audio**: SUCCESS
**localhost 'harmonicOscillator.js' Result**: SUCCESS
**localhost 'reverbTest.js' Result**: SUCCESS
**localhost 'playaPerformanceTest.js' Result**: SUCCESS
**dev-playa 'playaPerformanceTest.js' Result**: SUCCESS
**dev-demo 'playaPerformanceTest.js' Result**: N/A
**QA Pass**: SUCCESS

**Additional Test Results + Observations**:

In edit mode some of these entities ( dependent on extraction type ) render partially outside their bounding box.
This is more easily observed at large scales.

Some render with faces not showing. May be expected.


**Domain** |localhost |dev-playa |
---|---|---|---
**Render Rate**: |89 |89 |
**Present Rate**: |89 |89 |
**Present New Rate**: |90 |90 |
**CPU Buffer Count**: |2487 |3878 |
**GPU Buffer Count**: |330 |1141 |
**CPU Memory Usage**: |54990064 |69051600 |
**GPU Memory Usage**: |28472048 |42119920 |
**CPU Texture Count**: |230 |264 |
**GPU Texture Count**: |56 |213 |
**GPU Transfer Count**: |0 |0 |
**CPU Texture Memory Usage**: |517722332 |154858260 |
**GPU Texture Memory Usage**: |559209534 |1140295095 |
**GPU Virtual Memory Usage**: |559209534|1140295095|
**Triangle Count**: |362154 |2220975 |
**Triangle Rate**: |32923090 |222097500 |
**API Drawcall Count**: |67 |311 |
**Drawcall Count**: |122 |1090 |
**Drawcall Rate**: |11090 |109000 |
**State Changes Texture Count**:|77 |304 |
**Pipeline Count**: |34 |41 |
**Item Opaques**: |63 |469 |
**Item Translucents**: |20 |69 |
**Item Lights**: |4 |46 |
**Timing Opaques**: |0.263 |1.444 |
**Timing Translucents**: |0.033 |0.06 |
**Timing Lighting**: |0.022 |0.019 |
**Render Frame**: |1.037 |2.314 |

**Under bounds**
hifi-snap-by-piper-peppercorn-on-2017-03-20_15-05-05

**Beyond bounds**

hifi-snap-by-piper-peppercorn-on-2017-03-20_15-04-32

**Face normals ( expected )**

hifi-snap-by-piper-peppercorn-on-2017-03-20_13-17-44

File: libraries/gpu/src/gpu/Buffer.h [hunk 1]
```diff
@@ -198,7 +198,7 @@ class BufferView {
     BufferView(const BufferPointer& buffer, Size offset, Size size, const Element& element = DEFAULT_ELEMENT);
     BufferView(const BufferPointer& buffer, Size offset, Size size, uint16 stride, const Element& element = DEFAULT_ELEMENT);
 
-    Size getNumElements() const { return _size / _element.getSize(); }
+    Size getNumElements() const { return (_size - _offset) / _stride; }
```
Comment (AndrewMeadows) [inline libraries/gpu/src/gpu/Buffer.h]: @samcake, this change should interest you.  BTW: is it possible, either by accident or malicious intent, to create a buffer with zero `_stride`?

**Comment by sethalves** [comment]
yeah... that bounding-box thing isn't right, but I don't think it's related to this PR.  Thanks for the thorough check.

**Pull request merged**